### PR TITLE
Migrate DropIn to Braintree SDK v4

### PR DESCRIPTION
**Pull request by sshropshire**
### Summary of changes

 - Remove Braintree SDK v3 integration
 - Create `DropInClient` and use `BraintreeClient` internally for SDK calls

### Notes
- CI will fail for this branch until some changes to the core SDK are released. We have some in-flight changes in `braintree_android` that haven't been merged yet

### Checklist

 - [x] Added a changelog entry

- @sarahkoop
- @sshropshire

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,13 +2,24 @@
 
 ## unreleased
 
+* Add `DropInClient`
+* Add `DropInResultCallback`
+* Add `FetchMostRecentPaymentMethodCalback`
```
Comment (sestevens) [inline CHANGELOG.md:7]:
```suggestion
* Add `FetchMostRecentPaymentMethodCallback`
```

File: Demo/src/main/java/com/braintreepayments/demo/MainActivity.java [hunk 2]
```diff
@@ -242,29 +207,58 @@ protected void reset() {
...
+            public void onResult(DropInResult dropInResult, Exception error) {
+                if (dropInResult != null) {
+                    handleDropInResult(dropInResult);
+                } else {
+                    mAddPaymentMethodButton.setVisibility(VISIBLE);
+                }
+            }
+        });
+
+        dropInClient.deliverBrowserSwitchResult(this, new DropInResultCallback() {
```
Comment (sestevens) [inline Demo/src/main/java/com/braintreepayments/demo/MainActivity.java:238]: How come we're calling `deliverBrowserSwitchResult` in `onAuthorizationFetched`? I feel like I'm missing something here -- wouldn't the authorization be fetched well before any browser switching occurs?

File: Drop-In/src/main/java/com/braintreepayments/api/FetchMostRecentPaymentMethodCallback.java [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package com.braintreepayments.api;
+
+public interface FetchMostRecentPaymentMethodCallback {
+
+    void onResult(DropInResult dropInResult, Exception error);
```
Comment (sestevens) [inline Drop-In/src/main/java/com/braintreepayments/api/FetchMostRecentPaymentMethodCallback.java:5]: Should we have `@Nullable` on these params?

File: Demo/src/main/java/com/braintreepayments/demo/MainActivity.java (hunk 2 above)
Comment (sshropshire) [reply Demo/src/main/java/com/braintreepayments/demo/MainActivity.java:238]: Yeah the only reason is for process kill recovery. We'll probably look into caching the client token somehow, but we want to restructure the architecture a bit more first.

File: Drop-In/src/main/java/com/braintreepayments/api/FetchMostRecentPaymentMethodCallback.java (hunk 3 above)
Comment (sshropshire) [reply Drop-In/src/main/java/com/braintreepayments/api/FetchMostRecentPaymentMethodCallback.java:5]: Definitely yeah. We'll add these in.

File: Drop-In/src/main/java/com/braintreepayments/api/BaseActivity.java [hunk 4]
```diff
@@ -4,75 +4,46 @@
...
-        try {
-            mClientTokenPresent = Authorization.fromString(mDropInRequest.getAuthorization())
-                    instanceof ClientToken;
-        } catch (InvalidArgumentException e) {
-            mClientTokenPresent = false;
+    DropInClient getDropInClient() {
+        // lazily instantiate dropInClient for testing purposes
+        if (dropInClient != null) {
+            return dropInClient;
         }
```
Comment (scannillo) [inline Drop-In/src/main/java/com/braintreepayments/api/BaseActivity.java:38]: Why are we doing this instead of making DropInClient a property?

File: Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+
+    Authorization getAuthorization() {
+        return braintreeClient.getAuthorization();
+    }
+
+    void getConfiguration(ConfigurationCallback callback) {
+        braintreeClient.getConfiguration(callback);
+    }
+
+    void sendAnalyticsEvent(String eventFragment) {
```
Comment (scannillo) [inline Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java:84]: Should this be non-public / hidden from merchants?

File: Drop-In/src/main/java/com/braintreepayments/api/VaultedPaymentMethodSelectedListener.java [hunk 6]
```diff
@@ -0,0 +1,5 @@
+package com.braintreepayments.api;
+
+public interface VaultedPaymentMethodSelectedListener {
```
Comment (sestevens) [inline Drop-In/src/main/java/com/braintreepayments/api/VaultedPaymentMethodSelectedListener.java:3]: Should this be `VaultedPaymentMethodSelectedCallback` instead of `Listener`?

File: Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+
+        } else {
+            callback.onResult(false);
+        }
+    }
+
+    void tokenizePayPalRequest(FragmentActivity activity, PayPalFlowStartedCallback callback) {
+        PayPalRequest paypalRequest = dropInRequest.getPayPalRequest();
+        if (paypalRequest == null) {
+            paypalRequest = new PayPalVaultRequest();
```
Comment (scannillo) [inline Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java:137]: Should we instead just require they set the `paypalRequest` instead of defaulting to a vault request?

File: Drop-In/src/main/java/com/braintreepayments/api/DropInRequest.java [hunk 8]
```diff
@@ -68,7 +64,7 @@ public DropInRequest tokenizationKey(String tokenizationKey) {
     }
 
     /**
-     * Deprecated. Use {@link ThreeDSecureRequest#amount(String)}
+     * Deprecated. Use {@link ThreeDSecureRequest#setAmount(String)}
```
Comment (sestevens) [inline Drop-In/src/main/java/com/braintreepayments/api/DropInRequest.java:67]: I'm guessing y'all are leaving this for a separate PR, but we can/should remove this since it's deprecated.

File: Drop-In/src/main/java/com/braintreepayments/api/DropInRequest.java [hunk 9]
```diff
@@ -98,9 +94,9 @@ public DropInRequest collectDeviceData(boolean collectDeviceData) {
     /**
      * This method is optional.
      *
-     * @param request The Google Payment Request {@link GooglePaymentRequest} for the transaction.
+     * @param request The Google Payment Request {@link GooglePayRequest} for the transaction.
      */
-    public DropInRequest googlePaymentRequest(GooglePaymentRequest request) {
+    public DropInRequest googlePaymentRequest(GooglePayRequest request) {
```
Comment (sestevens) [inline Drop-In/src/main/java/com/braintreepayments/api/DropInRequest.java:99]: Is renaming this (and the setter) to use `googlePay` instead of `googlePayment` going to be part of a separate PR?

File: Drop-In/src/main/java/com/braintreepayments/api/SupportedPaymentMethodsAdapter.java [hunk 10]
```diff
@@ -10,59 +10,27 @@
...
-import java.util.Set;
+import java.util.List;
 
 class SupportedPaymentMethodsAdapter extends BaseAdapter {
 
-    private Context mContext;
-    private ArrayList<PaymentMethodType> mAvailablePaymentMethods;
-    private PaymentMethodSelectedListener mPaymentMethodSelectedListener;
+    private final List<DropInPaymentMethodType> supportedPaymentMethods;
+    private final PaymentMethodSelectedListener selectionListener;
```
Comment (sestevens) [inline Drop-In/src/main/java/com/braintreepayments/api/SupportedPaymentMethodsAdapter.java:19]:
```suggestion
    private final PaymentMethodSelectedListener selectedListener;
```

File: Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java [hunk 11]
```diff
@@ -0,0 +1,428 @@
...
+     * payment method, you may not need to show Drop-In.
+     * <p>
+     * Note: a client token must be used and will only return a payment method if it contains a
+     * customer id.
+     *
+     * @param activity the current {@link AppCompatActivity}
+     * @param callback callback for handling result
+     */
+    public void fetchMostRecentPaymentMethod(FragmentActivity activity, final FetchMostRecentPaymentMethodCallback callback) {
+        // TODO: consider sending back empty result if tokenization key auth
```
Comment (sestevens) [inline Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java:336]: I think throwing an error makes sense here. If merchants are calling this method with a tokenization key, they should find out right away that it won't work.

File: Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java [hunk 12]
```diff
@@ -0,0 +1,428 @@
...
+                    }
+                    callback.onResult(result, null);
+                } else if (error != null) {
+                    callback.onResult(null, error);
+                }
+            }
+        });
+    }
+
+    void getVaultedPaymentMethods(final FragmentActivity activity, boolean refetch, final GetPaymentMethodNoncesCallback callback) {
```
Comment (scannillo) [inline Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java:384]: Will all docstrings be a separate PR?

File: Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java (hunk 7 above)
Comment (sestevens) [reply Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java:137]: We default to a vault request in iOS Drop-in, so it makes sense to me for Android to be consistent with that.

https://github.com/braintree/braintree-ios-drop-in/blob/201f903ce6f444acb6676349e59eec3eb292f34b/Sources/BraintreeDropIn/BTPaymentSelectionViewController.m#L286

File: Drop-In/build.gradle [hunk 13]
```diff
@@ -58,9 +58,18 @@ android {
...
+    implementation 'androidx.annotation:annotation:1.1.0'
+
+    api 'com.braintreepayments.api:braintree-core:4.0.0-beta4-SNAPSHOT'
+    api 'com.braintreepayments.api:three-d-secure:4.0.0-beta4-SNAPSHOT'
+    api 'com.braintreepayments.api:paypal:4.0.0-beta4-SNAPSHOT'
+    api 'com.braintreepayments.api:venmo:4.0.0-beta4-SNAPSHOT'
+    api 'com.braintreepayments.api:google-pay:4.0.0-beta4-SNAPSHOT'
+    api 'com.braintreepayments.api:card:4.0.0-beta4-SNAPSHOT'
+    api 'com.braintreepayments.api:data-collector:4.0.0-beta4-SNAPSHOT'
+    api 'com.braintreepayments.api:union-pay:4.0.0-beta4-SNAPSHOT'
```
Comment (scannillo) [inline Drop-In/build.gradle:71]: We should update the CHANGELOG to include that these are all now included by default in DropIn.

File: build.gradle [hunk 14]
```diff
@@ -33,6 +33,7 @@ allprojects {
     repositories {
         google()
         jcenter()
+        mavenLocal()
```
Comment (scannillo) [inline build.gradle:36]: Does this need a TODO to be removed after testing?

File: Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java (hunk 5 above)
Comment (sarahkoop) [reply Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java:84]: This is package-privates so we should be good. It's used outside of this class so can't make it fully private.

File: Drop-In/src/main/java/com/braintreepayments/api/BaseActivity.java (hunk 4 above)
Comment (sshropshire) [reply Drop-In/src/main/java/com/braintreepayments/api/BaseActivity.java:38]: We inject a `DropInClient` into `DropInActivity` by overriding this method in the unit tests. We want to find a better way to inject mocks, but here we borrow the existing `BraintreeFragment` mocking pattern.

File: Drop-In/src/main/java/com/braintreepayments/api/DropInRequest.java (hunk 9 above)
Comment (sarahkoop) [reply Drop-In/src/main/java/com/braintreepayments/api/DropInRequest.java:99]: Yes we will do this in a separate PR - added a TODO for it

File: Drop-In/src/main/java/com/braintreepayments/api/SupportedPaymentMethodsAdapter.java (hunk 10 above)
Comment (sarahkoop) [reply Drop-In/src/main/java/com/braintreepayments/api/SupportedPaymentMethodsAdapter.java]: Renamed this 👍

File: Drop-In/src/main/java/com/braintreepayments/api/VaultedPaymentMethodSelectedListener.java (hunk 6 above)
Comment (sshropshire) [reply Drop-In/src/main/java/com/braintreepayments/api/VaultedPaymentMethodSelectedListener.java:3]: We decided to go with listener here since its invocation is directly tied to a `View.OnClickListener`.

File: Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java (hunk 12 above)
Comment (sarahkoop) [reply Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java:385]: Yes we want to wait to finalize the methods/API before we do docstrings. We added a TODO in `DropInClient` to add docstrings.

File: Drop-In/src/main/java/com/braintreepayments/api/DropInRequest.java (hunk 8 above)
Comment (sarahkoop) [reply Drop-In/src/main/java/com/braintreepayments/api/DropInRequest.java:67]: Yeah we will remove any deprecated methods in another PR 👍

File: Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java (hunk 11 above)
Comment (sarahkoop) [reply Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java]: Ok we will keep returning an error to the callback here then. Removed this TODO since it no longer applies.

File: Drop-In/build.gradle (hunk 13 above)
Comment (sarahkoop) [reply Drop-In/build.gradle:71]: Added this to the CHANGELOG

File: build.gradle (hunk 14 above)
Comment (sarahkoop) [reply build.gradle:37]: Added a TODO for this 👍

**Review by sestevens — APPROVED** [review]

File: Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java (hunk 5 above)
Comment (scannillo) [reply Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java:85]: Ah, the default is package-private. Thank you!

File: Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java (hunk 7 above)
Comment (scannillo) [reply Drop-In/src/main/java/com/braintreepayments/api/DropInClient.java:138]: Good eye 👍

**Review by scannillo — APPROVED** [review]
